cache/cache_test.go

```diff
@@ -0,0 +1,273 @@
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatalf("raw watcher saw no events – wrong range?")
+	}
+}
+
+func TestUnsupportedStartRev(t *testing.T) {
+	_, client := startEtcd(t)
+	ctx := context.Background()
+	c, _ := New(ctx, client, "")
+	defer c.Close()
+	if err := c.WaitReady(ctx); err != nil {
+		t.Fatal(err)
+	}
+
+	watchCh := c.Watch(ctx, "", clientv3.WithRev(123)) // any positive rev
+	resp, ok := <-watchCh
+	if !ok || !resp.Canceled {
+		t.Fatalf("expected immediate Canceled response for unsupported startRev, got %#v", resp)
+	}
+}
+
+// TODO: The following tests exercise non-zero startRev behavior.
+// TestTooOldRevisionCompact, TestWatchSurvivesServerCompaction
+// TestMaxOnceDelivery, TestReplayCompleteness
+// They’re currently removed until newWatchEventStream supports startRev>0.
+
+// helpers
+
+func startEtcd(t *testing.T) (*embed.Etcd, *clientv3.Client) {
+	t.Helper()
+
+	cfg := embed.NewConfig()
+	cfg.Dir = t.TempDir()
+	cfg.LogLevel = "error"
+
+	// pick free ports
+	newClient, _ := types.NewURLs([]string{"http://127.0.0.1:0"})
+	newPeer, _ := types.NewURLs([]string{"http://127.0.0.1:0"})
+	cfg.LCUrls, cfg.ACUrls = newClient, newClient
+	cfg.LPUrls, cfg.APUrls = newPeer, newPeer
+
+	// make the bootstrap string match the new peer URL
+	cfg.InitialCluster = fmt.Sprintf("%s=%s", cfg.Name, cfg.APUrls[0].String())
+
+	srv, err := embed.StartEtcd(cfg)
+	if err != nil {
+		t.Fatalf("start etcd: %v", err)
+	}
+	t.Cleanup(func() { srv.Close() })
+
+	select {
+	case <-srv.Server.ReadyNotify():
+	case <-time.After(10 * time.Second):
+		t.Fatalf("etcd ready timeout")
+	}
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{srv.Clients[0].Addr().String()},
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		t.Fatalf("new client: %v", err)
+	}
+	t.Cleanup(func() { client.Close() })
+	return srv, client
+}
+
+// waitEvent waits for a non‑empty WatchResponse up to 10s.
+func waitEvent(t *testing.T, ch clientv3.WatchChan, why string) clientv3.Event {
+	t.Helper()
+	deadline := time.After(10 * time.Second)
+	for {
+		select {
+		case watchResp, ok := <-ch:
+			if !ok {
+				t.Fatalf("%s: channel closed", why)
+			}
+			for _, ev := range watchResp.Events {
+				return *ev // return the first still-unread event
+			}
+		case <-deadline:
+			t.Fatalf("%s: timeout", why)
+		}
+	}
+}
+
+// waitUntil spins until pred() returns true or the timeout elapses.
+func waitUntil(t *testing.T, pred func() bool, timeout time.Duration) {
+	t.Helper()
+	deadline := time.After(timeout)
+	tick := time.NewTicker(5 * time.Millisecond)
+	defer tick.Stop()
+
+	for {
+		if pred() {
+			return
+		}
+		select {
+		case <-tick.C:
+		case <-deadline:
+			t.Fatalf("condition not met within %v", timeout)
+		}
+	}
+}
+
+// splitWatchResponses converts a WatchChan of batched responses into a WatchChan that emits one WatchResponse per individual Event.
+func splitWatchResponses(src clientv3.WatchChan) clientv3.WatchChan {
```

Comment:
This should not be needed, or it should be rewritten to just return a list of events.